### PR TITLE
[13.0][FIX] sale_order_type: Analytic account on copy

### PR DESCRIPTION
**Pull request by cesar-tecnativa**
@Tecnativa TT38012

### CODE REVIEW

**Comment by cesar-tecnativa** [comment]
@sergio-teruel please review

**Comment by pedrobaeza** [comment]
Incorrect commit and PR message. Follow the guidelines (https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#71commit-message)

- Short on main line
- Module name
- TT on the content, not on the main line

```
[FIX] sale_order_type: Analytic account on copy

When a sale order is duplicated, we set the default account analytic
account that is configured in the order type.

TT38012
```

**Comment by cesar-tecnativa** [comment]
Done

File: sale_order_type/models/sale.py [hunk 1]
```diff
@@ -125,6 +125,13 @@ def _prepare_invoice(self):
             res["sale_type_id"] = self.type_id.id
         return res
 
+    def copy(self, default=None):
+        if self.type_id and self.type_id.analytic_account_id:
+            default = dict(
+                default or {}, analytic_account_id=self.type_id.analytic_account_id.id,
+            )
+        return super().copy(default)
```
Comment (pedrobaeza) [inline sale_order_type/models/sale.py]: You should read the type on the defaults, as you don't know if there's any override or default value set on the DB (and respect kw args):

```suggestion
        res = super().copy(default=default)
        sale_type = self.env["sale.order.type"].browse(res.get("type_id", False)) or self.type_id
        if sale_type.analytic_account_id:
            res["analytic_account_id"] = sale_type.analytic_account_id.id
        return res
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by sergio-teruel** [comment]
@cesar-tecnativa Can you add some test to coverage this.?

**Comment by cesar-tecnativa** [comment]
> @cesar-tecnativa Can you add some test to coverage this.?

done

File: sale_order_type/models/sale.py [hunk 2]
```diff
@@ -125,6 +125,13 @@ def _prepare_invoice(self):
             res["sale_type_id"] = self.type_id.id
         return res
 
+    def copy(self, default=None):
```
Comment (pedrobaeza) [inline sale_order_type/models/sale.py:163]: Doing this in the method `copy_data` you avoid an extra write, although you would need to do a browse over ` type_id` ID result if exists.

File: sale_order_type/tests/test_sale_order_type.py [hunk 3]
```diff
@@ -218,3 +231,15 @@ def test_sale_order_quotation_validity(self):
         order = self.create_sale_order()
         order.onchange_type_id()
         self.assertEqual(fields.Date.to_string(order.validity_date), "2022-01-11")
+
+    def test_sale_copy_function(self):
+        """
+        Test when duplicating the sale order the account analytic account is set.
+        """
```
Comment (pedrobaeza) [inline sale_order_type/tests/test_sale_order_type.py:314]: PEP257 indicates that the docstring should be put this way:

```suggestion
        """Test when duplicating the sale order the account analytic account is set.
        """
```

File: sale_order_type/tests/test_sale_order_type.py [hunk 4]
```diff
@@ -50,6 +50,14 @@ def setUp(self):
         self.immediate_payment = self.env.ref("account.account_payment_term_immediate")
         self.sale_pricelist = self.env.ref("product.list0")
         self.free_carrier = self.env.ref("account.incoterm_FCA")
+        self.analytic_account = self.env["account.analytic.account"].create(
```
Comment (pedrobaeza) [inline sale_order_type/tests/test_sale_order_type.py:54]: You should put these changes in a separate commit with the message:

```
[IMP] sale_order_type: Don't depend on demo data for tests
```

for not mixing unrelated things in the diff.

**Comment by pedrobaeza** [comment]
@carlosdauden please attend my comments.

**Comment by pedrobaeza** [comment]
@sergio-teruel can you please do the requested changes?

**Comment by rousseldenis** [comment]
@sergio-teruel What's the status of this ?

**Pull request closed without merging**